I6102 ms word cross ref links

feerrenrut opened the pull request:
Add support for cross references in word. Fixes #6102 

Now able to report cross references as links when reading through text in word. Can browse by link and when in browse mode a Cross ref link can be activated.

Had to expand the range to the paragraph because single words with in a `ref` dont report as refs. Only the first and the last character.

Refs that are links have the `fields.item(1).code.text` property set to something like `u' REF _Ref457210120 \\h`

For performance reasons the links are cached first, then as we traverse over the words in the range we check if the word is in the vector of known links.

feerrenrut committed:
Add support for cross references in word

now able to report cross references as links when reading through text in word.
Browse mode, navigating by links with 'k' does not work
Activating the links do not work.
Probably need more validation to check that the ref is also a link. This
can be done by checking for the presence of \\h in the
`fields.item(1).code.text` result eg: `u' REF _Ref457210120 \\h `

feerrenrut committed:
Cross ref in MSWord can be activated

When in browse mode a Cross ref link can be activated. This currently
has some limitations. If there are multiple words in the field, internal
words can not be used to jump to the bookmark.

feerrenrut committed:
Handle navigating to links in browse mode

feerrenrut committed:
Fix xRef links not reporting mid range

cross reference links now report as links on characters / words between
the start and end. The links can be activated with the caret at
(immediately before) the first character.

feerrenrut committed:
Converted #defines to constexpr

For type safety.

feerrenrut committed:
Removed unnecessary parameter

Other calls to expand only pass in one parameter, also the argument
string passed only declares one parameter.

feerrenrut committed:
Convert link functions to return bool

Since we only ever check if there is more than 0, this allows them to
exit early.

feerrenrut committed:
Improve performance by caching the link locations

Link locations are cached before entering a loop that steps
through the range word by word.

feerrenrut commented:
@michaelDCurran would you mind taking a look at this when you have the chance?

michaelDCurran commented on source/NVDAObjects/window/winword.py:
```diff
@@ -437,8 +445,18 @@ def iterate(self):
+		if t == FIELD_TYPE_REF:
+			fieldText = item.code.text.strip().split(' ')
+			# ensure that the text has a \\h in it
+			return any( fieldText[i] == '\\h' for i in range(2, len(fieldText)) )
```
range should be xrange

michaelDCurran commented on source/NVDAObjects/window/winword.py:
```diff
@@ -499,6 +517,32 @@ def activate(self):
+			tiCopy = self.copy()
+			tiCopy.expand(textInfos.UNIT_LINE)
+			speech.speakTextInfo(tiCopy,reason=controlTypes.REASON_FOCUS)
+			braille.handler.handleCaretMove(tiCopy)
```
Should be tempRange (I.e. the collapsed one)

michaelDCurran commented on source/NVDAObjects/window/winword.py:
```diff
@@ -499,6 +517,32 @@ def activate(self):
+			# text will be something like ' REF _Ref457210120 \\h '
+			fieldText = field.code.text.strip().split(' ')
+			# the \\h field indicates that the field is a link
+			if not any( fieldText[i] == '\\h' for i in range(2, len(fieldText)) ):
```
xrange not range also.

feerrenrut committed:
Review actions for #6242

Fix sillyness in the inRange function.
convert range to xrange in python code.
convert format flags to hex
pull out addition of link xml tag from `generateXMLAttribsForFormatting`
fix issue allowing links to be activated one character before they start

michaelDCurran commented:
All looks good to me.

feerrenrut committed:
Add #include guards
